### PR TITLE
Fetch membergroup lists instead of members

### LINKED ISSUE #6200: Notification Memory Exhaustion Bug
#### Description
My forum has 1.7 million members currently.  I know I can fix this error by increasing the maximum php memory size but I'd rather not have to do that if I can help it.

### Steps to reproduce
1. Go to your profile
2. Click Forum Profile -> Notifications -> Notifications Preferences

### Environment (complete as necessary)
- Version/Git revision: 2.1 Latest Github version
- Database Type: mysql
- Database Version: mysql  Ver 15.1 Distrib 10.2.19-MariaDB, for Linux (x86_64) using readline 5.1
- PHP Version: 7.4

### Additional information/references
image

### PR DESCRIPTION
**Pull request by live627**
membersAllowedTo() already calls groupsAllowedTo(),  so don't extrapolate members from groups. That query we bypass is also really slow.

Fixes #6200

### CODE REVIEW

**Comment by Butterflysaway** [comment]
This fix is very important.  I've personally checked that this fixes the issue entirely.  Please merge.

File: Sources/Subs-Membergroups.php [hunk 1]
```diff
@@ -830,4 +830,70 @@ function list_getMembergroups($start, $items_per_page, $sort, $membergroup_type)
...
+ * @param int   $profile_id
+ *
+ * @return array An array containing two arrays - 'allowed', which has which groups are allowed to do it and 'denied' which has the groups that are denied
+ */
+function getGroupsWithPermissions(array $group_permissions = array(), array $board_permissions = array(), $profile_id = 1)
+{
+	global $smcFunc;
+
+	$member_groups = array();
+	if (!empty($group_permissions))
```
Comment (MissAllSunday) [inline Sources/Subs-Membergroups.php:847]: You can avoid unneeded nesting by checking for empty($group_permissions) and return early

Comment (live627) [reply Sources/Subs-Membergroups.php:847]: No

**Comment by live627** [comment]
```
PHPUnit 8.5.2 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.10
Configuration: C:\Users\John\repos\SMF2.1\phpunit.xml
Error:         No code coverage driver is available

..                                                                  2 / 2 (100%)

Time: 171 ms, Memory: 8.00 MB

OK (2 tests, 25 assertions)
```

**Pull request merged**